### DISCUSSION TITLE
Should `setRelatedReferences` stay?

### DISCUSSION THREAD

**Discussion by foundrytom** [OpenAssetIO/OpenAssetIO#18] [Ideas and change proposals]
A question was raised as to "what are the use cases for [`ManagerInterface.setRelatedReferences`](https://github.com/TheFoundryVisionmongers/OpenAssetIO/blob/14dc134703fd9b3431c933bdb992e85bd67c8dce/python/openassetio/managerAPI/ManagerInterface.py#L974)?". This really asks the question, is this a mutation something that we should have in the API?

The original design of this API is to provide a two-way abstraction between a host, and a manager.  For the manager, it (in theory) enables asset publishing from any OAIO host without specific implementation. For a host, it allows first-class asset-based workflows to be developed without having to know the specifics of any given asset management system.

`getRelatedReferences` is to allow a querying of dependencies at runtime. Eg. get me the shots for this sequence, or get me all the child assets for this city.

 `setRelatedReferences` (though see #16) is to allow the host to register relationships it may know about in a generic way. Eg: after publishing a new asset (eg: a render), adding the document that produced that asset (eg. nuke/katana script) as a relationship to the render via the appropriate `RelationshipSpecifciation` (eg. `SourceDocumentRelationship`).

As the API is also designed as a general abstraction of a _tool_ and an asset management system, not just for a third-party DCC, it was always envisioned that the API might be used by other parts of the pipeline that may well have a strong reason to want to define relationships.

What do people think? Does it make sense to keep this symmetry in the API, and encourage hosts to implement out-the-box functionality such as those above?

**Comment by j-stroud** [comment]
This would be a common use case for us. As you mention, you might want to add additional relationships between assets after a publish.
If the API is to support publishing new assets or versions of assets then this functionality would be required. The alternative to providing an explicit method for this would be if the publishing API itself allowed this information to be provided at the point of publish - though we've also had cases were you are publishing a set of interrelated assets, you really need to publish the entities first, then construct a graph of relationships later.
There's a host of separate issues related to publishing that one would need to consider but are probably best left to an implementation and discussion elsewhere - atomic transactions, eventual consistency etc.

**Reply by pshinners-blz** [reply]
I'd argue against this, but wouldn't complain if it existed. If a developer is writing tools for our facility, I would expect them to use the Asset API and libraries supported by our studio. No need for them to abstract through OAIO.

It feels like the potential value for this call is to allow generic tools or plugins that could be shared across studios. But this sounds dubious, and I wouldn't expect much alignment between studio naming conventions and workflows enough to generically define relationships between assetIds. A strong counter point could be studios making use of Katana's existing "attribute" and "relationship" calls.

I'll share example code from our current FnAssetPlugin. I'd prefer to not even need these; looking back on `getAssetAttributes` I find this dubious to provide.
```

void Example::getRelatedAssetId(const std::string& assetId, const std::string& relation, std::string& ret)
{
    ret = "";
}

void Example::setAssetAttributes(const std::string& assetId, const std::string& scope, const StringMap& attrs)
{
}

void Example::getAssetAttributes(const std::string& assetId, const std::string& scope, StringMap& returnAttrs)
{
    if (ignorevalue(assetId)) {
        return;
    }
    auto entity = LookupEntity(assetId);
    if (!entity.IsNull()) {
        for (auto& iter: entity.GetObject()) {
            if (iter.value.IsString()) {
                returnAttrs[iter.name.GetString()] = iter.value.GetString();
            }
        }
    }
}

```

**Reply by foundrytom** [reply]
> It feels like the potential value for this call is to allow generic tools or plugins that could be shared across studios. But this sounds dubious, and I wouldn't expect much alignment between studio naming conventions and workflows enough to generically define relationships between assetIds. 

When researching this API a while back, a common task that people wanted  post-render (batch) hooks for was to associate the script that produced a render with the rendered asset. One goal of OAIO is to make facility integration of off-the-shelf tools easier. We envisaged that defining a relationship specification in the standard API spec library (eg: `SourceDocument` or similar), and having the DCC automatically add that relationship to any assetized renders would help reduce the need for host API specific code in the pipeline integration. Ie Katana, Nuke, (Houdini, Maya...) would all use the same well-known relationship spec (from `openassetio-cgi.specifications`), and provide the reference to the source document for you, after publishing the render. This would hopefully mean you don't have to go fishing through their respective APIs. The first-class, well known specifications in the common API hopefully avoids ambiguity about which spec to use. A manager could chose to ignore relationships added with this spec if they wanted, but its there if they cared.

Another common case was also to associate a list of concrete asset versions used in the render - something that a host could also provide as out-the-box functionality if desired.  

This is why there we added a `setRelated` as well as ` get`. 

I can certainly see how in-house development would only merit from this mechanism if there was a high-level mandate to decouple code from a specific asset manager where possible. This sounds potentially utopian, but I'd guess the tradeoffs are similar to OpenCL vs CUDA?

So if we look at this through the lens of how can this help ISVs write tools that are easier to integrate into production pipelines - is there value in the above use of `setRelatedReferences`?
 
> I'll share example code from our current FnAssetPlugin. I'd prefer to not even need these; looking back on getAssetAttributes I find this dubious to provide.

There is a very different meaning of a call similar to that in OAIO, which we can delve into separately. This hopefully brings far greater utility as it was provided to solve some frequently cited integration challenges, and helps with one of the many reasons facilities have needed to replace Read/Write nodes in DCC tools.

> If the API is to support publishing new assets or versions

Fwiw, In the API, `setRelatedReferences` is never intended to be used to manage versions. This is left up to the manager to decide during `register`, a topic we need to extend the documentation on, which we will def do soon. There is a note to this ambiguity, but it doesn't have enough detail.

Many thanks for all the input both, most helpful.
